shared-libs/transitions/src/transitions/registration.js

```diff
@@ -75,6 +72,181 @@ const booleanExpressionFails = (doc, expr) => {
   return result;
 };
 
+const getDOB = doc => {
+  if (!doc || !doc.fields) {
+    return '';
+  }
+  const reportedDate = moment(doc.reported_date).startOf('day');
+  const years = parseInt(getYearsSinceDOB(doc), 10);
+  if (!isNaN(years)) {
+    return reportedDate.subtract(years, 'years');
+  }
+  const months = parseInt(getMonthsSinceDOB(doc), 10);
+  if (!isNaN(months)) {
+    return reportedDate.subtract(months, 'months');
+  }
+  const weeks = parseInt(getWeeksSinceDOB(doc), 10);
+  if (!isNaN(weeks)) {
+    return reportedDate.subtract(weeks, 'weeks');
+  }
+  const days = parseInt(getDaysSinceDOB(doc), 10);
+  if (!isNaN(days)) {
+    return reportedDate.subtract(days, 'days');
+  }
+  // no given date of birth - return reportedDate as it's the best we can do
+  return reportedDate;
+};
+
+const getYearsSinceDOB = doc => {
+  const fields = ['years_since_dob', 'years_since_birth', 'age_in_years'];
+  return findFirstDefinedValue(doc, fields);
+};
+
+const getMonthsSinceDOB = doc => {
+  const fields = ['months_since_dob', 'months_since_birth', 'age_in_months'];
+  return findFirstDefinedValue(doc, fields);
+};
+
+const getWeeksSinceDOB = doc => {
+  const fields = [
+    'weeks_since_dob',
+    'dob',
+    'weeks_since_birth',
+    'age_in_weeks',
+  ];
+  return findFirstDefinedValue(doc, fields);
+};
+
+const getDaysSinceDOB = doc => {
+  const fields = ['days_since_dob', 'days_since_birth', 'age_in_days'];
+  return findFirstDefinedValue(doc, fields);
+};
+
+/*
+ * Given a doc get the LMP value as a number, including 0.  Supports three
+ * property names atm.
+ * */
+const getWeeksSinceLMP = doc => {
+  const props = ['weeks_since_lmp', 'last_menstrual_period', 'lmp'];
+  const prop = props.find(prop => !isNaN(Number(doc.fields && doc.fields[prop])));
+  return Number(prop && doc.fields[prop]);
+};
+
+const setExpectedBirthDate = doc => {
+  const lmp = Number(getWeeksSinceLMP(doc)),
+        start = moment(doc.reported_date).startOf('day');
+  if (lmp === 0) {
+    // means baby was already born, chw just wants a registration.
+    doc.lmp_date = null;
+    doc.expected_date = null;
+  } else {
+    start.subtract(lmp, 'weeks');
+    doc.lmp_date = start.toISOString();
+    doc.expected_date = start
+      .clone()
+      .add(40, 'weeks')
+      .toISOString();
+  }
+};
```

Comment:
I just noticed this... what happens if the lmp is undefined or otherwise NaN because the field isn't defined or invalid? The old code was the same way so I wonder if we check for that earlier in the stack but it might be worth adding a check here so it does nothing rather than trying to use it. What do you think?

Reply:
I've noticed this too. 
If `lmp` is undefined or NaN, the `doc.lmp_date` ends up to be  the reported_date. (`moment.subtract(NaN, 'whatever')` doesn't change the date). 
The thing is ... this part of the code only gets called if the `add_expected_date` trigger is set. I would think that we should error if no `lmp` is set, but I'm afraid that would break someone's configuration.

Reply:
Yeah, it's quite possible that some people rely on it defaulting to now. I guess it's fine for now...